NativeAuthSampleApp/MultiFactorAuthenticationViewController.swift

```diff
@@ -0,0 +1,329 @@
+
+    var verifyCodeViewController: VerifyCodeViewController?
+
+    var accountResult: MSALNativeAuthUserAccountResult?
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        do {
+            nativeAuth = try MSALNativeAuthPublicClientApplication(
+                clientId: Configuration.clientId,
+                tenantSubdomain: Configuration.tenantSubdomain,
+                challengeTypes: [.OOB, .password]
+            )
+        } catch {
+            print("Unable to initialize MSAL \(error)")
+            showResultText("Unable to initialize MSAL")
+        }
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        retrieveCachedAccount()
+    }
+
+    @IBAction func signInPressed(_: Any) {
+        guard let email = emailTextField.text, let password = passwordTextField.text else {
+            resultTextView.text = "Email or password not set"
+            return
+        }
+
+        print("Signing in with email \(email) and password")
+
+        showResultText("Signing in...")
+
+        nativeAuth.signIn(username: email, password: password, delegate: self)
+    }
+    
+    @IBAction func signOutPressed(_: Any) {
+        guard accountResult != nil else {
+            print("signOutPressed: Not currently signed in")
+            return
+        }
+        accountResult?.signOut()
+
+        accountResult = nil
+
+        showResultText("Signed out")
+
+        updateUI()
+    }
+
+    func showResultText(_ text: String) {
+        resultTextView.text = text
+    }
+
+    func updateUI() {
+        let signedIn = (accountResult != nil)
+
+        signInButton.isEnabled = !signedIn
+        signOutButton.isEnabled = signedIn
+    }
+
+    func retrieveCachedAccount() {
+        accountResult = nativeAuth.getNativeAuthUserAccount()
+        if let accountResult = accountResult, let homeAccountId = accountResult.account.homeAccountId?.identifier {
+            print("Account found in cache: \(homeAccountId)")
+
+            accountResult.getAccessToken(delegate: self)
+        } else {
+            print("No account found in cache")
+
+            accountResult = nil
+
+            showResultText("")
+
+            updateUI()
+        }
+    }
+}
+
+// MARK: - Sign In delegates
+
+// MARK: SignInStartDelegate
+
+extension MultiFactorAuthenticationViewController: SignInStartDelegate {
+
+    func onSignInStartError(error: MSAL.SignInStartError) {
+        print("SignInStartDelegate: onSignInStartError: \(error)")
+
+        if error.isUserNotFound || error.isInvalidCredentials || error.isInvalidUsername {
+            showResultText("Invalid username or password")
+        } else {
+            showResultText("Error while signing in: \(error.errorDescription ?? "No error description")")
+        }
+    }
+
+    func onSignInAwaitingMFA(newState: AwaitingMFAState) {
+        print("SignInStartDelegate: onSignInAwaitingMFA: \(newState)")
```

Comment:
If I'm not wrong like this, we're printing the object index of "newState". I don't think it's useful

Reply:
Done